src/OpenTelemetry/Internal/CircularBufferStruct.cs

```diff
@@ -174,17 +166,25 @@ public T Read()
             var index = (int)(this.tail % this.Capacity);
             while (true)
             {
-                T value = this.trait[index];
-                if (value == null)
+                if (!this.trait[index].IsReady)
                 {
                     // If we got here it means a writer isn't done.
                     continue;
                 }
 
-                this.trait[index] = null;
+                // TODO: we are doing an extra copy from the buffer, this can be optimized if Read() could take a callback
```

Comment:
I wonder if it would help to use ref return here? Something like this...

```csharp
public ref T Read()
{
   var value = ref this.trait[index].Value;
   this.trait[index].Value = default(T);
   this.trait[index].IsReady = false;
   this.tail++;
   return ref value;
}
```

If T is readonly struct, compiler might be smart enough to do that automatically.

For max perf this might also work...
```csharp
public ref T Read()
{
   ref var value = ref this.trait[index].Value;
   // this.trait[index].Value = default(T); <- Not sure if this will clobber the ref or not.
   this.trait[index].IsReady = false;
   this.tail++;
   return ref value;
}
```

Reply:
Once `this.tail++` happened, other writer threads have the freedom to overwrite the struct value, which will cause race condition if the reader hasn't finished consuming the struct?

Reply:
TBH, not sure. This is fringe stuff I only rarely get a chance to use 😄 I had a mistake in the first snippet, updated now. I removed a ref that shouldn't have been in there. The first snippet I think is safe. It makes a copy from the array, and then returns that copy by ref to the caller. At best, saves one copy. Second snippet might have a race. There are strict rules with ref locals, caller has to also use it as a local. So it will be extremely short-lived, but yes there could be an issue if the buffer loops around really quickly. Maybe we go as-is and then we can try to perf hack it while you continue on with the real effort?

Reply:
I bet the 1st one is covered by NVR (named return value) optimization 😄